### PR TITLE
Issue 1765: MultiReaderTxnWriterWithFailoverTest txn ping

### LINKED ISSUE #1765: MultiReaderTxnWriterWithFailoverTest needs to ping txns
**Problem description**
We have observed in one run that txns are timing out and the controller is rightfully aborting them. Here is a sample of the log messages showing the problem:

```
2017-08-19 17:27:13,540 87059 [segment-store-42] DEBUG i.p.s.s.impl.hdfs.CreateOperation - 
Creating 
testMultiReaderWriterTxnScope370693410/testMultiReaderWriterTxnStream/1#transaction.4589ceb6
304440aab224f29fc3eb07ad

2017-08-19 17:27:19,078 930254 [controllerpool-60] DEBUG i.p.c.t.TimerWheelTimeoutService - 
Successfully executed abort on tx 
testMultiReaderWriterTxnScope370693410/testMultiReaderWriterTxnStream/4589ceb6-3044-40aa-
b224-f29fc3eb07ad

2017-08-19 17:27:19,089 92608 [epollEventLoopGroup-3-7] DEBUG 
i.p.s.s.c.StreamSegmentContainer - SegmentContainer[1]: deleteStreamSegment 
[testMultiReaderWriterTxnScope370693410/testMultiReaderWriter
TxnStream/1#transaction.4589ceb6304440aab224f29fc3eb07ad]

E7-08-19 17:27:25,643 99162 [epollEventLoopGroup-3-4] ERROR i.p.s.s.h.handler.AppendProcessor - 
Error (Segment = 
'testMultiReaderWriterTxnScope370693410/testMultiReaderWriterTxnStream/1#transaction.4
589ceb6304440aab224f29fc3eb07ad', Operation = 'append')
io.pravega.segmentstore.contracts.StreamSegmentNotExistsException: [Segment 
'testMultiReaderWriterTxnScope370693410/testMultiReaderWriterTxnStream/1#transaction.4589ceb6
304440aab224f29fc3eb07ad'] The
 StreamSegment does not exist.

```

**Problem location**
System tests

**Suggestions for an improvement**
Ping created txns.

### PR DESCRIPTION
**Pull request by fpj**
**Change log description**
* Adds logic to ping txns while they are open

**Purpose of the change**
Fixes #1765 

**What the code does**
It uses the the utility methods in `FutureHelpers` to ping open txns frequently.

**How to verify it**
Run system tests.

### CODE REVIEW

File: test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java [hunk 1]
```diff
@@ -406,8 +427,8 @@ private void performFailoverTest() {
                     txnStatusFutureList.add(checkTxnStatus(transaction, eventWriteCount));
                 } catch (Throwable e) {
                     log.warn("Exception while writing events in the transaction: {}", e);
-                    if (transaction != null) {
-                        log.debug("Transaction with id: {}  failed", transaction.getTxnId());
+                    if (txnDebugReference != null) {
+                        log.debug("Transaction with id: {}  failed", txnDebugReference.getTxnId());
```
Comment (shrids) [inline test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java]: Q: Should we cancel the the future returned by FutureHelpers.loop incase of an exception? `java.util.concurrent.CompletableFuture#cancel`

File: test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java [hunk 2]
```diff
@@ -387,12 +388,32 @@ private void performFailoverTest() {
...
+                    Transaction<Long> transaction = retry
                             .retryingOn(MultiReaderTxnWriterWithFailoverTest.TxnCreationFailedException.class)
                             .throwingOn(RuntimeException.class)
                             .run(() -> createTransaction(writer, stopWriteFlag));
+                    txnDebugReference = transaction;
+                    AtomicBoolean txnIsDone = new AtomicBoolean(false);
+
+                    // Sets a recurrent delayed task to ping the txn. It exits when the
+                    // txn completes and no longer needs pinging
+                    FutureHelpers.loop(txnIsDone::get, () -> {
```
Comment (shiveshr) [inline test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java]: we can use cancellable request to achieve this.

File: test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java [hunk 3]
```diff
@@ -387,12 +388,32 @@ private void performFailoverTest() {
...
+                                    transaction.ping(5000);
+                                    return null;
+                                }, new CompletableFuture<Void>());
+                            } else {
+                                txnIsDone.set(true);
+                            }
+
+                            return null;
+                        }, Duration.ofMillis(2000), executorService);
+                    }, executorService);
```
Comment (shrids) [inline test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java]: Minor : we need to change line 421 from `transaction` to `transaction.getTxnId()`
https://github.com/pravega/pravega/pull/1766/files#diff-fd4ef3523065620e53c0447293088fd0R421

File: test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java [hunk 4]
```diff
@@ -387,12 +388,32 @@ private void performFailoverTest() {
...
                             .throwingOn(RuntimeException.class)
                             .run(() -> createTransaction(writer, stopWriteFlag));
+                    txnDebugReference = transaction;
+                    AtomicBoolean txnIsDone = new AtomicBoolean(false);
+
+                    // Sets a recurrent delayed task to ping the txn. It exits when the
+                    // txn completes and no longer needs pinging
+                    FutureHelpers.loop(txnIsDone::get, () -> {
+                        return FutureHelpers.delayedTask(() -> {
+                            if (transaction.checkStatus() == Transaction.Status.OPEN) {
```
Comment (Prabha-Veerubhotla) [inline test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java]: Minor Query: should `txnDebugReference`  be used here? and for `ping` ?

File: test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java [hunk 5]
```diff
@@ -172,7 +173,7 @@ public void setup() {
         assertTrue(segmentStoreInstance.isRunning());
         log.info("Pravega Segmentstore service instance details: {}", segmentStoreInstance.getServiceDetails());
         //executor service
-        executorService = Executors.newScheduledThreadPool(NUM_READERS + NUM_WRITERS);
+        executorService = Executors.newScheduledThreadPool(NUM_READERS + NUM_WRITERS + 2);
```
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java]: Not related to this PR, but I don't find this too scalable. Do we really need a dedicated thread for each reader and each writer?

File: test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java (hunk 3 above)
Comment (fpj) [inline test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java]: done

File: test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java (hunk 4 above)
Comment (fpj) [inline test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java]: no, the compiler doesn't like it that way, that's why I had to create another reference.

File: test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java (hunk 5 above)
Comment (shrids) [reply test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java]: Having a separate thread for each of the readers and writers is helpful for debugging tests. This was helpful to find deadlock /thread block issues.

**Pull request merged**